app/controllers/user_badges_controller.rb

```diff
@@ -109,10 +109,13 @@ def toggle_favorite
       return render json: failed_json, status: 400
     end
 
+    new_is_favorite_value = !user_badge.is_favorite
     UserBadge
       .where(user_id: user_badge.user_id, badge_id: user_badge.badge_id)
```

Comment:
It's always smart to select the values you want to update by adding `where(is_favorite: !new_is_favorite_value)`. It seems counter intuitive but it performs better in psql.

Reply:
Hmm, a good point in general, but I'm afraid in this specific case it can cause false assumptions that a part of users' badges of _specific type_ can be marked as favorite and a part is not. Which isn't the case, we always toggle _all_ users' badges of _specific type_.

Reply:
I don't quite understand this, could you expand and perhaps provide an example? I am suggesting the query should go from something like this:

```sql
UPDATE user_badges SET is_favorite = false WHERE user_id = 1234 and badge_id = 3457
```

To:

```sql
UPDATE user_badges SET is_favorite = false WHERE user_id = 1234 and badge_id = 3457 AND is_favorite = true
```

The query is the same expect `is_favorite` is added to the where with the opposite value of what we are updating it to, to prevent updating records that don't need to be updated.

I do not understand how it would create a false assumption.

Reply:
> to prevent updating records that don't need to be updated.

This is the false assumption ⬆️. 

There couldn't exist a record that doesn't need to be updated 🙂 If everything works as expected, this query can't return badges with different `is_favorite` values:
```sql
SELECT * 
FROM user_badges 
WHERE user_id = 1234 and badge_id = 3457
```
Either all of them have `is_favorite == true`, either all of them have `is_favorite == false`. Here is an example, user has 16 "God topic" badges:
<img width="343" alt="Screenshot 2021-07-15 at 23 39 40" src="https://user-images.githubusercontent.com/1274517/125847607-84a5c705-9e56-49e5-ae6c-b034ce9eab07.png">

It's 16 rows in a database table. If user choose this badge as favorite all 16 rows will have `is_favorite == true`, if user remove this badge from favorites all 16 rows will have `is_favorite == false`. It's not ideal implementation on the db level, we probably need a separate table. But this is what we have now.

As I understand, in this specific case adding an additional condition won't speed up anything but can cause confusion.
Or maybe there is something I don't see?

Reply:
In general, I understand that adding such condition even though looks counterintuitive may work faster in database.

Reply:
I prefer to always add a condition like that because it's a good habit to build up and I don't personally find it confusing, but I'll leave this decision with you since in this particular case it doesn't seem needed. No change is required.